### ISSUE TITLE
Persistent L2ARC invalid ashift

### ISSUE THREAD

**Issue by atkinsam** [openzfs/zfs#14313]
### System information
Type | Version/Name
 --- | ---
Distribution Name | Amazon Linux 2
Distribution Version	| Amazon Linux 2 AMI 2.0.20221210.1 x86_64
Kernel Version	| 5.10.157-139.675.amzn2.x86_64
Architecture	| x86_64
OpenZFS Version	| 2.1.4 - 2.1.7

### Describe the problem you're observing
I am observing an issue concerning persistent L2ARC. I use a non-default ashift for my zpool and an L2ARC cache with the default ashift (i.e. I don't provide -o ashift=X when adding my cache device). When I reboot after populating the L2ARC, ZFS fails to validate the checksum of the L2ARC header on my device and the L2ARC rebuild fails. Any subsequent reboots of the host result in a successful L2ARC rebuild. Only the rebuild on the first reboot after adding the L2ARC cache device fails. 

I originally thought that this could be the same issue as https://github.com/openzfs/zfs/issues/11787, but I don't see any "failed to unpack label" errors when running zdb -lll on my cache device. I don't see any evidence of any real corruption. I'm also using a gpt partition as my cache device, which is called out as a solution in the same issue. Leading me to create this separate issue. 

### Describe how to reproduce the problem
1. Create a zpool with ashift=12
2. Add an L2ARC cache device to the pool. I'm using a gpt partition on an nvme disk as my device.
3. Populate the L2ARC with some data
4. Reboot
5. L2ARC header checksum fails, L2ARC rebuild fails
6. Populate the L2ARC with some more data, reboot again, and L2ARC rebuild succeeds

### Include any warning/errors/backtraces from the system logs

Without any custom patches, logging, etc. I see the following on the first reboot after adding the L2ARC cache device when the rebuild fails:

```
[root@ip-198-19-64-195 ~]# cat /proc/spl/kstat/zfs/dbgmsg | grep "L2"
1670454195   arc.c:10403:l2arc_dev_hdr_read(): L2ARC IO error (52) while reading device header, vdev guid: 6934761350446009203
1670454195   arc.c:10403:l2arc_dev_hdr_read(): L2ARC IO error (52) while reading device header, vdev guid: 6934761350446009203
1670454226   spa_history.c:307:spa_history_log_sync(): txg 801 L2ARC rebuild no valid log blocks
```

When I add logging from this commit (https://github.com/atkinsam/zfs/commit/275026ae4a6b5c1da503b02e71d8ef81fcaaa312), I can see that the ashift value for the L2ARC device is set to 12 (the ashift for my zpool) when I first add the L2ARC device:

```
[root@ip-198-19-95-57 ~]# cat /proc/spl/kstat/zfs/dbgmsg | grep "L2"
1671719412   spa.c:1951:spa_load_l2cache(): L2ARC ashift 12, asize 0
1671719412   arc.c:9944:l2arc_add_vdev(): Using asize 4096 for L2ARC vdev
```

And on the first reboot after adding the L2ARC device, I can see that the ashift value is determined to be the default of 9, and the L2ARC rebuild fails:

```
[root@ip-198-19-95-57 ~]# cat /proc/spl/kstat/zfs/dbgmsg | grep "L2"
1671720950   spa.c:1951:spa_load_l2cache(): L2ARC ashift 0, asize 0
1671720950   arc.c:9944:l2arc_add_vdev(): Using asize 512 for L2ARC vdev
1671720950   arc.c:10433:l2arc_dev_hdr_read(): L2ARC IO error (52) while reading device header, vdev guid: 1421654579314785763
```

When I add another change, from this commit (https://github.com/atkinsam/zfs/commit/faa9afa06cd574c297cf7529603dea44497351f8), the L2ARC is able to successfully rebuild on the initial reboot after adding the L2ARC:

```
### after adding the L2ARC device
[root@ip-198-19-95-57 ~]# cat /proc/spl/kstat/zfs/dbgmsg | grep "L2"
1671722648   spa.c:1951:spa_load_l2cache(): L2ARC ashift 12, asize 0
1671722648   spa.c:1954:spa_load_l2cache(): Overwrite L2ARC ashift
1671722648   arc.c:9944:l2arc_add_vdev(): Using asize 512 for L2ARC vdev

### after initial reboot
[root@ip-198-19-95-57 ~]# cat /proc/spl/kstat/zfs/dbgmsg | grep "L2"
1671723144   spa.c:1951:spa_load_l2cache(): L2ARC ashift 0, asize 0
1671723144   spa.c:1954:spa_load_l2cache(): Overwrite L2ARC ashift
1671723144   arc.c:9944:l2arc_add_vdev(): Using asize 512 for L2ARC vdev
1671723176   spa_history.c:307:spa_history_log_sync(): txg 2445 L2ARC rebuild successful, restored 3 blocks
```

Obviously my commit is not a good long-term solution. I'm wondering if others have thoughts on how this should be solved. Maybe I'm missing something, but I can't find the place where the L2ARC ashift value is stored. It doesn't seem to be in the nvlist for the L2ARC device. Is the intention for the L2ARC device to inherit the ashift value from the spa?

**Comment by adamdmoss** [comment]
FAO @gamanakis perhaps!

**Comment by behlendorf** [comment]
Nice job running this down.  It does seem like the issue is that the ashift isn't stored for the L2ARC vdev since prior to the persistent ARC feature it wasn't needed.  One possible option would be to write out a more complete label for persistent L2ARC devices.  @gamanakis @grwilson may have some thoughts about this.

**Comment by gamanakis** [comment]
Nice catch! As far as I remember L2ARC devices don't have normal labels, instead the have a special header. I could incorporate the ashift information there.

**Comment by allanjude** [comment]
So, ZoL made ashift this "kind-of" property. Do we want to consider making it an actual property, possibly via vdev properties, and persist the value to the vdev's ZAP, so that we can get the correct value always?

**Comment by atkinsam** [comment]
Thanks all for taking a look so quickly. Is it possible to read the header on the device to determine the ashift... before the ashift is known? I might be missing something. My understanding was that the read of the header itself was failing because the ashift was wrong.

**Comment by clhedrick** [comment]
If I understand correctly, you could get the rebuild to work if you initially specified an ashift of 9 when adding the cache, right? It looks to me like the initial reboot fails because by default ashift=12 is taken from the pool for the initial add. But at reboot, it defaults to 9. That means the cache is recreated with 9. At that point successive reboots work because that 9 agrees with the (erroneous) default of 9 used during reboot. Of course this isn't very friendly to the SSD.

A workaround that wouldn't require changes on disk would be to save the 12 in the cache file.

**Comment by gamanakis** [comment]
Playing around with it, we could just store the ashift in the label of the device, and this solves the problem.
A primitve draft:
```
diff --git a/module/zfs/vdev_label.c b/module/zfs/vdev_label.c
index 85c7134ca..ca604e66c 100644
--- a/module/zfs/vdev_label.c
+++ b/module/zfs/vdev_label.c
@@ -486,7 +486,7 @@ vdev_config_generate(spa_t *spa, vdev_t *vd, boolean_t getstats,
        if (vd->vdev_isspare)
                fnvlist_add_uint64(nv, ZPOOL_CONFIG_IS_SPARE, 1);

-       if (!(flags & (VDEV_CONFIG_SPARE | VDEV_CONFIG_L2CACHE)) &&
+       if (!(flags & VDEV_CONFIG_SPARE) &&
            vd == vd->vdev_top) {
                fnvlist_add_uint64(nv, ZPOOL_CONFIG_METASLAB_ARRAY,
                    vd->vdev_ms_array);
```

**Comment by shodanshok** [comment]
@gamanakis awesome find, thanks. One question: I have an `ashift=12` pool with a persistent L2ARC device, but I was only [hit once by the issue](https://github.com/openzfs/zfs/issues/11787#issuecomment-1380962243). Is the problem expected to be consistently reproducible or I had another issue entirely?

**Comment by gamanakis** [comment]
That particular issue occurs only once, at the first import of the pool, if the pool has an ashift other than 9. I think you should be good with your current cache device even without the patch.
